### DISCUSSION TITLE
How do we figure out what "broke" the expression?

### DISCUSSION THREAD

**Discussion by rchougule** [antonmedv/expr#711] [General]
Hello,

Let's say we have a complex long expression which evaluates to a boolean value. How do we figure out which part "broke" it? i.e. which part led the evaluation to result into `false`?

For e.g.
```
let result = Func1() && Var1 == "X" && Func2(Var2) && (Var3 in [x,y,z] || var4 in [a,b,c]);

return result.Value, result.Reason
```

The above expression might give some idea on the expectation. The `result.Reason` can be `Func2(Var2) is false` (let's say).

Thinking out loud here. Do you think this is achievable? I understand that what `reason` means for one might not mean for someone else. Though, we can try to make it deterministic in some way.

But do you think this is possible?

Thanks,
Rohan Chougule

**Comment by antonmedv** [comment]
You can use Explain mode from [Expr Pro](https://expr-lang.org/docs/expr-pro/explain):
<img width="740" alt="image" src="https://github.com/user-attachments/assets/c98fb411-d2ab-4bdf-9129-aefd82f1c53a">
This way, it is possible to see exactly why expression "broke"

**Reply by rchougule** [reply]
Oh, it's a paid feature. Got it.
Though, how will it exactly work via code? Any examples? Will I get a list of these "steps" when I invoke `expr.Run`?

**Reply by antonmedv** [reply]
Yes, exactly. You provide code as a string and env object. On output - list of steps as strings.
